connman/gdbus/client.c

```diff
@@ -853,28 +853,30 @@ gboolean g_dbus_proxy_method_call(GDBusProxy *proxy, const char *method,
 	if (client == NULL)
 		return FALSE;
 
-	data = g_try_new0(struct method_call_data, 1);
-	if (data == NULL)
-		return FALSE;
-
-	data->function = function;
-	data->user_data = user_data;
-	data->destroy = destroy;
-
 	msg = dbus_message_new_method_call(client->service_name,
 				proxy->obj_path, proxy->interface, method);
-	if (msg == NULL) {
-		g_free(data);
+	if (msg == NULL)
 		return FALSE;
-	}
 
 	if (setup) {
 		DBusMessageIter iter;
 
 		dbus_message_iter_init_append(msg, &iter);
-		setup(&iter, data->user_data);
+		setup(&iter, user_data);
 	}
 
+	if (!function)
+		return g_dbus_send_message(client->dbus_conn, msg);
+
+	data = g_try_new0(struct method_call_data, 1);
+	if (data == NULL)
+		return FALSE;
+
+	data->function = function;
+	data->user_data = user_data;
+	data->destroy = destroy;
```

Comment:
random observation: looks like  the idea is that caller yields ownership of user_data and expects that destroy callback will be called for it at some point. And there already were N ways for that not to happen (every mid-function return) => potential memory leaks. And now there is a new one: the freshly added shortcut for "we do not care about return message so lets skip user data cleanup too" case... what was the point of this commit again?

Reply:
I think this was made because someone in their plugin(s) did not want to pass a function to the `g_dbus_proxy_method_call()` since only thing using this is the Bluetooth plugin.

Reply:
Ok, later on this is more extensively used by `iwd` plugin and systemd DNS resolver.

Reply:
But was [3b0a203](https://github.com/sailfishos/connman/pull/66/commits/3b0a2036ecfa0647b90cf602ad19319a1c907d17) something you were after?

Reply:
It has been a while, so IIRC: yeah, the function had minor problems to start with, and now  after some changes it has new ones including possibility of becoming a memory leak generator even on a supposedly happy path.